### PR TITLE
Pin depends

### PR DESCRIPTION
**Pull request by TheLortex**
This is a rebase of @rizo's PR: https://github.com/ocamllabs/opam-monorepo/pull/159

Fixes https://github.com/ocamllabs/opam-monorepo/issues/153

### CODE REVIEW

File: cli/lock.ml [hunk 1]
```diff
@@ -109,6 +109,43 @@ let root_depexts local_opam_files =
...
+let root_pin_depends local_opam_files =
+  OpamPackage.Name.Map.fold
+    (fun _pkg (_version, opam_file) acc -> OpamFile.OPAM.pin_depends opam_file @ acc)
+    local_opam_files []
+
+let pull_pin_depends (pin_depends : (OpamPackage.t * OpamUrl.t) list) =
+  (* TODO: Group pin-depends with the same url. *)
+  if List.is_empty pin_depends then Ok OpamPackage.Name.Map.empty
+  else
+    let pins_tmp_dir = Fpath.v "/tmp/opam-monorepo/pins/" in
```
Comment (emillon) [inline cli/lock.ml:121]: Can we use a tempdir function like `Bos.OS.Dir.with_tmp` here?

File: cli/lock.ml [hunk 2]
```diff
@@ -109,6 +109,43 @@ let root_depexts local_opam_files =
     (fun _pkg (_version, opam_file) acc -> OpamFile.OPAM.depexts opam_file :: acc)
     local_opam_files []
 
+let root_pin_depends local_opam_files =
+  OpamPackage.Name.Map.fold
+    (fun _pkg (_version, opam_file) acc -> OpamFile.OPAM.pin_depends opam_file @ acc)
+    local_opam_files []
+
+let pull_pin_depends (pin_depends : (OpamPackage.t * OpamUrl.t) list) =
+  (* TODO: Group pin-depends with the same url. *)
```
Comment (emillon) [inline cli/lock.ml:118]: Can this be taken care of or a ticket created for this?

**Comment by emillon** [comment]
I just rebased and added a changelog entry.

**Comment by NathanReb** [comment]
Thanks for the multiple rebase, I'll revive this PR.

There are a couple things that I'd like to improve, in particular the interface for `Opam_solve` and code sharing for pull and pulling the pin depends.

File: cli/lock.ml (hunk 1 above)
Comment (NathanReb) [reply cli/lock.ml]: I implemented this bit!

File: cli/lock.ml (hunk 2 above)
Comment (NathanReb) [reply cli/lock.ml]: I implemented it and removed the comment!

**Comment by NathanReb** [comment]
I made a few improvements:
- Clean up the interface of `Opam_solve` and let it deal with pin_depends rather than merging them into `local_opam_files` in `Lock`. `calculate` now expects a map for local packages and a map for pin-depends, abstracting everything else away from `Lock`.
- I moved the pin_depends extraction further down the chain so that we only load opam states once
- I fixed a small quirk in `Opam_solve.get_opam_info` so that the pin-depends map is only looked up once instead of twice (`find_opt` instead of `mem` + `find`)
- I forbid having local packages in a pin-depends field. Before `lock` would have allowed it but the pin would effectively be ignored since local packages are removed from the solution in the lockfile. There's no way to properly support this as you either want it local and use whatever you have in your repo or use a pin that will be pulled in your duniverse. Having both will lead to a dune error because you have the same package twice in your workspace.
- I factored opam pulling code for pin-depends and the pull command
- I added deduplication of the repository's pin-depends (the same pin could appear several times in different opam files)
- I added rejection of different pins for the same package
- we group pin-depends per URL, meaning if several packages are pinned to the same URLs, the repo is pulled once instead of once per package
- I implemented @emillon suggestion to use `Bos` to generate a temporary directory

I need to test this a bit.
@TheLortex would you mind trying this out for mirage and see if it still works?
@emillon could you give this another round of review?

**Review by emillon — APPROVED** [review]

**Pull request merged**